### DISCUSSION TITLE
Unable to install matplotlib, please help

### DISCUSSION THREAD

**Discussion by hexise** [termux/termux-app#3444] [Q&A] [answered]
Hello,

I want to run some Python code in Termux, I want to install matplotlib in Termux.

When I install matplotlib, I have permission denied errors. I am using the Termux version 0.118.0 on Galaxy S21 phone with Android 13 installed. I also grant storage permission to Termux.

The execution log is as following, I do not know what I need to do next. When I install matplotlib on my another ubuntu testing machine, there is no problem.

`~ $ pkg install matplotlib
Checking availability of current mirror:
[*] https://mirror.vern.cc/termux/termux-main: ok
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  matplotlib
0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded.
Need to get 0 B/6533 kB of archives.
After this operation, 28.9 MB of additional disk space will be used.
Selecting previously unselected package matplotlib.
(Reading database ... 21940 files and directories currently installed.)
Preparing to unpack .../matplotlib_3.7.1_aarch64.deb ...
Unpacking matplotlib (3.7.1) ...
Setting up matplotlib (3.7.1) ...
Installing dependencies through pip. This may take a while...
Collecting contourpy>=1.0.1
  Using cached contourpy-1.0.7.tar.gz (13.4 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done            Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Preparing metadata (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [15 lines of output]
      /data/data/com.termux/files/usr/tmp/pip-build-env-12aa4j6r/overlay/lib/python3.11/site-packages/setuptools/config/pyprojecttoml.py:66: _BetaConfiguration: Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
        config = read_configuration(filepath, True, ignore_option_errors, dist)
      running dist_info
      creating /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info
      writing /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/PKG-INFO
      writing dependency_links to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/dependency_links.txt
      writing requirements to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/requires.txt
      writing top-level names to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/top_level.txt
      writing manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt'
      reading manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt'
      reading manifest template 'MANIFEST.in'
      adding license file 'LICENSE'
      writing manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt'
      creating '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy-1.0.7.dist-info'
      error: [('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/PKG-INFO', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/PKG-INFO', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/PKG-INFO'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/not-zip-safe', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/not-zip-safe', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/not-zip-safe'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/dependency_links.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/dependency_links.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/dependency_links.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/requires.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/requires.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/requires.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/top_level.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/top_level.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/top_level.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/SOURCES.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__/SOURCES.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.__bkp__'")]
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
dpkg: error processing package matplotlib (--configure):
 installed matplotlib package post-installation script subprocess returned error exit status 1
Errors were encountered while processing:
 matplotlib
E: Sub-process /data/data/com.termux/files/usr/bin/dpkg returned an error code (1)`

**Comment by hexise** [comment]
Some more information I found.

If I run the same Termux command on an Android 10 device, matplotlib can be installed successfully.

Is there anything I need to do on Android 13 device? Is it a known issue for Termux on Android 13? Thanks.

I have run pkg install build-essential on both devices before installing module.

**Comment by sylirre** [comment, marked as answer]
You need the latest `python` and `python-pip` packages. Do this:
* `apt purge matplotlib`
* `apt update && apt dist-upgrade -yq`
* `apt install -yq matplotlib`

These permission denials were because Pip attempted to set extended attributes on files which is not allowed on Android. Issue should be fixed with newer package versions.

**Edit: see answers under https://github.com/termux/termux-app/discussions/3444#discussioncomment-6394303**

**Reply by patham9** [reply]
Your "answer" does not work unfortunately, due to the issue mentioned below regarding ninja.
I guess the Termux project has a hard time chasing the ever-changing ever-breaking Python dependencies, which also includes very commonly used ones like Matplotlib.

**Reply by dieuhai1234** [reply]
- [ ] termux_maps

**Comment by techmarvy** [comment]
I tried the solution mentioned by @sylirre , But does not work for me . 
Python is  and python-pip are latest 
I have installed numpy  :  `pkg install python-numpy `  , it works and install numpy 1.25.0 which is released recently. 
But `apt install -yq matplotlib` show error as below, Is it because of updated numpy version. 

```
 ERROR: Failed building wheel for ninja
      Failed to build ninja
      ERROR: Could not build wheels for ninja, which is required to install pyproject.toml-based projects
```

**Reply by sylirre** [reply]
Unlikely this is due to numpy version. The failed ninja module (dependency) has nothing to do with numpy.

**Comment by techmarvy** [comment]
Hmm, any idea, how can I resolve this . I can install python-numpy but matplotlib installation does not work .

**Reply by patham9** [reply]
This got broken very recently, I had Matplotlib working on Termux one month ago.

**Reply by techmarvy** [reply]
Exactly same here , it was working well until last month . Then there were few dependencies got updates like `numpy1.25.0` released on June 17th,2023
As if i use 
`pkg install matplotlib`   or `pkg install python-numpy` .
It  always installs the latest version available, with this matplotlib have some issues and does not install .
I changed my python script finally, instead of using matplotlib to plot the graph , i am using plotly .
Which can be installed via `pip install plotly`.
Please share if someone get matplotlib working again on termux.

**Comment by Vladik01-11** [comment]
There was the same problem, the video from YouTube helped), by entering all the commands, Matplotlib and Pandas were installed (but I don't really understand this anymore). The bottom line is that it helped, I hope it will help you too
https://youtu.be/mPy3XoRTByM

**Reply by TermuxMonet** [reply]
Thats the only solution that worked! Thanks!

Here's the commands for anyone else who also needs help.

`CFLAGS=-Wno-implicit-function-declaration MATHLIB=m`
`pip install meson`
`pip install meson-python`
`pip install pybind11`
`pip install --no-build-isolation contourpy`
`pkg install ninja automake cmake binutils patchelf`
`pkg install matplotlib`

**Reply by Yahia-Ghadiry** [reply]
This resolved it(wheel error for contourpy ninja dependency)

**Reply by Woodenonez** [reply]
Finally find the way out. Thank you!

**Reply by patham9** [reply]
These instructions are a bit half-*ssed unfortunately as a few crucial steps are missing from a fresh Termux installation. I will try to fix this when I have time.

**Reply by apchhi** [reply]
Thenx bro. Problem solved

**Comment by Mohammad-Azizi** [comment]
It is easy to fix it you just need to allow file access permissions to termux by going to mobile settings then apps then find termux and look for permissions click on it and you will see permissions those aren't allowed

**Reply by sylirre** [reply]
File access permissions do not have effect on what happens inside Termux. They only affect users access to `/storage/emulated/0` (shared storage). Granting this permission will not resolve issue with Matplotlib, as problem origin not comes from there.

Installation of packages or Python modules not require access to shared storage as nothing is going to be written on it.

**Comment by KrishAnjara** [comment]
Try Installing ninja and patchelf independently for supported versions

**Comment by Alekc74** [comment]
> Hello,
> 
> I want to run some Python code in Termux, I want to install matplotlib in Termux.
> 
> When I install matplotlib, I have permission denied errors. I am using the Termux version 0.118.0 on Galaxy S21 phone with Android 13 installed. I also grant storage permission to Termux.
> 
> The execution log is as following, I do not know what I need to do next. When I install matplotlib on my another ubuntu testing machine, there is no problem.
> 
> `~ $ pkg install matplotlib Checking availability of current mirror: [*] https://mirror.vern.cc/termux/termux-main: ok Reading package lists... Done Building dependency tree... Done Reading state information... Done The following NEW packages will be installed: matplotlib 0 upgraded, 1 newly installed, 0 to remove and 0 not upgraded. Need to get 0 B/6533 kB of archives. After this operation, 28.9 MB of additional disk space will be used. Selecting previously unselected package matplotlib. (Reading database ... 21940 files and directories currently installed.) Preparing to unpack .../matplotlib_3.7.1_aarch64.deb ... Unpacking matplotlib (3.7.1) ... Setting up matplotlib (3.7.1) ... Installing dependencies through pip. This may take a while... Collecting contourpy>=1.0.1 Using cached contourpy-1.0.7.tar.gz (13.4 MB) Installing build dependencies ... done Getting requirements to build wheel ... done Installing backend dependencies ... done Preparing metadata (pyproject.toml) ... error error: subprocess-exited-with-error
> 
> × Preparing metadata (pyproject.toml) did not run successfully. │ exit code: 1 ╰─> [15 lines of output] /data/data/com.termux/files/usr/tmp/pip-build-env-12aa4j6r/overlay/lib/python3.11/site-packages/setuptools/config/pyprojecttoml.py:66: _BetaConfiguration: Support for `[tool.setuptools]` in `pyproject.toml` is still _beta_. config = read_configuration(filepath, True, ignore_option_errors, dist) running dist_info creating /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info writing /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/PKG-INFO writing dependency_links to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/dependency_links.txt writing requirements to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/requires.txt writing top-level names to /data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/top_level.txt writing manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt' reading manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt' reading manifest template 'MANIFEST.in' adding license file 'LICENSE' writing manifest file '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt' creating '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy-1.0.7.dist-info' error: [('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/PKG-INFO', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/PKG-INFO', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/PKG-INFO'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/not-zip-safe', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/not-zip-safe', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/not-zip-safe'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/dependency_links.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/dependency_links.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/dependency_links.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/requires.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/requires.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/requires.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/top_level.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/top_level.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/top_level.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info/SOURCES.txt', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/SOURCES.txt', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**/SOURCES.txt'"), ('/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info', '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**', "[Errno 13] Permission denied: '/data/data/com.termux/files/usr/tmp/pip-modern-metadata-aoixkupr/contourpy.egg-info.**bkp**'")] [end of output]
> 
> note: This error originates from a subprocess, and is likely not a problem with pip. error: metadata-generation-failed
> 
> × Encountered error while generating package metadata. ╰─> See above for output.
> 
> note: This is an issue with the package mentioned above, not pip. hint: See above for details. dpkg: error processing package matplotlib (--configure): installed matplotlib package post-installation script subprocess returned error exit status 1 Errors were encountered while processing: matplotlib E: Sub-process /data/data/com.termux/files/usr/bin/dpkg returned an error code (1)`

curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py

python3 get-pip.py

pkg i tur-repo

pkg i ninja patchelf openblas gcc-11

pip install virtualenv

python3 -m venv venv

source venv/bin/activate

pip install yq meson-python

ln -sf $PREFIX/bin/ld.gold $PREFIX/bin/ld

pkg install fftw libzmq freetype

pip install matplotlib==3.7.3

deactivate

pip install meson

pip install yq meson-python

CFLAGS=-Wno-implicit-function-declaration MATHLIB=m pip install numpy==1.23.2

pip install matplotlib==3.7.3
